### PR TITLE
add memory limit for aggregations

### PR DESCRIPTION
**Pull request by PSeitz**
introduce AggregationLimits to set memory consumption limit and bucket limits
memory limit is checked during aggregation, bucket limit is checked before returning the aggregation request.

TODO: add check for term aggregation

### CODE REVIEW

File: src/aggregation/agg_limits.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+/// An estimate for memory consumption
+pub trait MemoryConsumption {
+    fn memory_consumption(&self) -> usize;
+}
+
+impl<K, V, S> MemoryConsumption for HashMap<K, V, S> {
+    fn memory_consumption(&self) -> usize {
+        let num_items = self.len();
+        (std::mem::size_of::<K>() + std::mem::size_of::<V>()) * num_items
```
Comment (fulmicoton) [inline src/aggregation/agg_limits.rs]:
```suggestion
        let num_items = self.capacity();
        (std::mem::size_of::<K>() + std::mem::size_of::<V>()) * num_items
```

File: src/aggregation/agg_limits.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+use super::collector::DEFAULT_MEMORY_LIMIT;
+use super::{AggregationError, DEFAULT_BUCKET_LIMIT};
+use crate::TantivyError;
+
+/// An estimate for memory consumption
+pub trait MemoryConsumption {
+    fn memory_consumption(&self) -> usize;
+}
+
+impl<K, V, S> MemoryConsumption for HashMap<K, V, S> {
```
Comment (fulmicoton) [inline src/aggregation/agg_limits.rs:16]:
```suggestion
impl<K: Copy, V: Copy, S> MemoryConsumption for HashMap<K, V, S> {
```

File: src/aggregation/error.rs [hunk 3]
```diff
@@ -3,7 +3,29 @@ use super::bucket::DateHistogramParseError;
 /// Error that may occur when opening a directory
 #[derive(Debug, Clone, PartialEq, Eq, Error)]
 pub enum AggregationError {
-    /// Failed to open the directory.
+    /// Date histogram parse error
     #[error("Date histogram parse error: {0:?}")]
     DateHistogramParseError(#[from] DateHistogramParseError),
+    /// Memory limit exceeded
+    #[error(
+        "Aborting aggregation because memory limit was exceeded. Limit: {limit:?}, Current: \
```
Comment (fulmicoton) [inline src/aggregation/error.rs:13]: that would be nice to display a value expressed in MB, with a unit.

File: src/error.rs [hunk 4]
```diff
@@ -55,7 +55,7 @@ impl fmt::Debug for DataCorruption {
 #[derive(Debug, Clone, Error)]
 pub enum TantivyError {
     /// Error when handling aggregations.
-    #[error("AggregationError {0:?}")]
+    #[error(transparent)]
```
Comment (fulmicoton) [inline src/error.rs:58]: ah cool ! I did not know about `transparent`

File: src/aggregation/bucket/histogram/histogram.rs [hunk 5]
```diff
@@ -287,6 +296,12 @@ impl SegmentAggregationCollector for SegmentHistogramCollector {
 }
 
 impl SegmentHistogramCollector {
+    fn get_memory_consumption(&self) -> usize {
```
Comment (fulmicoton) [inline src/aggregation/bucket/histogram/histogram.rs:299]: we should really add byte-unit as a dependency of tantivy at one point... if you agree can you open a ticket?

File: src/aggregation/bucket/histogram/histogram.rs [hunk 6]
```diff
@@ -287,6 +296,12 @@ impl SegmentAggregationCollector for SegmentHistogramCollector {
...
+        let self_mem = std::mem::size_of::<Self>();
+        let sub_aggs = self.sub_aggregations.memory_consumption();
+        let buckets = self.buckets.memory_consumption();
+        self_mem + sub_aggs + buckets
```
Comment (fulmicoton) [inline src/aggregation/bucket/histogram/histogram.rs]:
```suggestion
        let self_mem = std::mem::size_of::<Self>();
        let sub_aggs_mem = self.sub_aggregations.memory_consumption();
        let buckets_mem = self.buckets.memory_consumption();
        self_mem + sub_aggs_mem + buckets_mem
```

**Review by fulmicoton — APPROVED** [review]

File: src/aggregation/error.rs (hunk 3 above)
Comment (PSeitz) [reply src/aggregation/error.rs:13]: I changed the existing `ByteCount` to cover that

**Pull request merged**